File: paper_artifacts/experiments/llr8/artifacts/sources/oss120b-fortran/tsvc_2_s231__submitted.f90
! Optimized Fortran implementation of TSVC kernel s231
! Exposes C entry point: tsvc_2_s231_fp64
subroutine tsvc_2_s231_fp64(aa, bb, LEN_2D, workspace, ws_bytes) bind(C, name="tsvc_2_s231_fp64")
    use iso_c_binding
    implicit none
    !--- scalar arguments (value) ---
    integer(c_int64_t), value :: LEN_2D
    type(c_ptr), value :: workspace
    integer(c_int64_t), value :: ws_bytes
    !--- array arguments ---
    real(c_double), dimension(*), intent(inout) :: aa
    real(c_double), dimension(*), intent(in) :: bb
    !--- locals ---
    integer :: i, j
    integer(c_int64_t) :: idx

    ! Parallel outer loop over columns (i)
    !$omp parallel do schedule(static) default(none) private(i, j, idx) &
    !$omp& shared(aa, bb, LEN_2D)
    do i = 1, LEN_2D
                        do j = 2, LEN_2D
            idx = (j - 1) * LEN_2D + i
            aa(idx) = aa(idx - LEN_2D) + bb(idx)
        end do
    end do
    !$omp end parallel do

    return
end subroutine tsvc_2_s231_fp64

File: paper_artifacts/experiments/llr8/artifacts/sources/oss120b-c-skills/tsvc_2_s231__submitted.c
/*
 * TSVC kernel "s231" implementation in C.
 *
 * The reference Python implementation (see /shared/tasks/tsvc_2_s231/tsvc_2_s231_numpy.py)
 * performs a column-wise prefix sum:
 *
 *   for i in range(LEN_2D):
 *       for j in range(1, LEN_2D):
 *           aa[j,i] = aa[j-1,i] + bb[j,i]
 *
 * In C the arrays are stored row-major, so the element aa[j,i] is located at
 *   aa[j*LEN_2D + i]
 * and similarly for bb. Each column (fixed i) is independent, therefore the
 * outer loop over i can be parallelised with OpenMP. The inner loop carries a
 * true data dependence on the previous row, so it must remain sequential.
 */

#include <stdint.h>
#include <omp.h>

/*
 * Function signature follows the benchmark convention: the function name matches
 * the kernel key (the last path component) and the arguments are plain pointers
 * to the data followed by the size parameter. The arrays contain double-precision
 * floating‑point values. `restrict` informs the compiler that the pointers do
 * not alias, enabling better optimisation.
 */
void tsvc_2_s231_fp64(double *restrict aa, const double *restrict bb, int64_t LEN_2D, uint8_t *workspace, int64_t workspace_bytes)
{
    /* Parallelise over columns (index i). Each thread works on a contiguous
       span of columns, providing good cache locality and avoiding false
       sharing. */
    #pragma omp parallel for schedule(static)
    for (int64_t i = 0; i < LEN_2D; ++i) {
        /* The first row (j = 0) is left unchanged – it already contains the
           initial value required for the recurrence. */
        for (int64_t j = 1; j < LEN_2D; ++j) {
            /* Compute linear indices for the current and previous rows.
               The stride between rows is LEN_2D because the innermost dimension
               (i) varies fastest in C's row-major layout. */
            aa[j * LEN_2D + i] = aa[(j - 1) * LEN_2D + i] + bb[j * LEN_2D + i];
        }
    }
}
